Metadata Editor and Widget Bug Fixes

ajbozarth opened the pull request:
Various Bug Fixes as a followup to #591

- Fixed required validation in MetadataEditor. Update to strict
equality check cause the `undefined` case to no longer be caught
by the `null` check
- Text updates to RuntimesWidget to clarify rutime type and api url
- Fixed support for RuntimeWidget when no runtimes are configured
- Focus on Name field when opening a Metadata Editor tab
- Update pipeline submission success dialog to use `display_name`

Fixes #798
Fixes #802
Fixes #806
Fixes #814 


 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

lresende commented on packages/metadata-common/src/MetadataEditor.tsx:
```diff
@@ -123,7 +123,8 @@ export class MetadataEditor extends ReactWidget {
-        (this.metadata[schemaField] === null ||
+        (this.metadata[schemaField] === undefined ||
+          this.metadata[schemaField] === null ||
           this.metadata[schemaField] === '' ||
```
What if you do something like  `&& (this.metadata[schemaField] || this.metadata[schemaField] === '(No selection)')` where the first comparison should take care of null, undefined and empty string

kevin-bates replied:
How about a local function `this.is_field_set(schemaField: str): boolean`?  Seems like there are way too many variations to deal with.

ajbozarth replied:
@lresende we can't do that because 0 or false may be valid and are also considered falsy when evaluating `this.metadata[schemaField]`, also `[]` is considered truthy. I do like the idea of a separate function like @kevin-bates suggested, I'll do that.

ajbozarth replied:
pushed an update

kevin-bates commented on packages/metadata-common/src/MetadataEditor.tsx:
```diff
@@ -110,6 +110,16 @@ export class MetadataEditor extends ReactWidget {
+      schemaField === [] ||
+      schemaField === '(No selection)'
+    );
+  }
```
The way this is implemented, it's not necessary to be a class method but works.  I figured it would act on `this.metadata` and take the field name as a parameter.  

If we could at least change the parameter name to `schemaValue` then that might eliminate some potential confusion.  But then the question arises - should this be named `IsValueEmpty()`?  :smile:

ajbozarth replied:
this would be a good function to put in a util, but I didn't think it was worth creating a util.ts just for this function (thus making it a private class method).

As for naming, you are correct, based on my current implementation it should be value not field. I'll push an update promptly

ptitzler commented:
@ajbozart just noticed a regression. The tool tip used to be correct.


![image](https://user-images.githubusercontent.com/13068832/89294567-f8a4ab80-d65f-11ea-882b-e47e7255526d.png)

ajbozarth commented:
Pushed an update that addressed the following issues @ptitzler raised:

- Focus on Name field when opening a Metadata Editor (including switching between tabs)
- Fixed race condition causing `Create new undefined` tooltip
- Fixed #814

karlaspuldaro commented on packages/metadata-common/src/MetadataEditor.tsx:
```diff
@@ -110,6 +114,16 @@ export class MetadataEditor extends ReactWidget {
 
+  private isValueEmpty(schemaValue: any): boolean {
+    return (
+      schemaValue === undefined ||
+      schemaValue === null ||
+      schemaValue === '' ||
```
We can just use `!schemaValue` instead of these 3 value checks

ajbozarth replied:
I mentioned this in a review comment from Luciano but we can't use that because it would evaluate 0 and false as falsy as well, but those might be valid content.

karlaspuldaro commented on packages/metadata-common/src/MetadataEditor.tsx:
```diff
@@ -110,6 +114,16 @@ export class MetadataEditor extends ReactWidget {
+      schemaValue === undefined ||
+      schemaValue === null ||
+      schemaValue === '' ||
+      schemaValue === [] ||
```
This is probably not gonna work since an array is an object, if we do [] === [] we get `false`
To be safe, we could do `Array.isArray(schemaValue) && schemaValue.length`

ajbozarth replied:
nice catch, I'll make that change

marthacryan commented:
Tried locally and it worked as expected for all of the issues mentioned. LGTM
